File: SessionCharacter.cs
using System.Buffers.Text;

namespace BlackTower;

public class SessionCharacter
{
	public static readonly SessionCharacter DefaultCharacter = new(Character.DefaultCharacter);

	public Character Character { get; }

	public SessionCharacter(Character character)
	{
		this.Character = character;
	}
	public JobClass? EquippedClass { get; set; } = null;

	public uint CurrentHP { get; set; } = 0;
	public uint CurrentMP { get; set; } = 0;
	public byte MaxHP => (byte)(Character.BaseHP + this.Character.SpecialtyClass.BaseHP + (this.EquippedClass?.BaseHP ?? 0) +
								(this.Character.MainHandWeapon?.HP ?? 0) + (this.Character.OffHandWeapon?.HP ?? 0) +
								(this.Character.Armor?.HP ?? 0) + (this.Character.Accessory?.HP ?? 0));

	public byte MaxMP => (byte)(Character.BaseMP + this.Character.SpecialtyClass.BaseMP + (this.EquippedClass?.BaseMP ?? 0));

	public byte Strength =>
		(byte)(Character.BaseStrength + this.Character.SpecialtyClass.BaseStrength + (this.EquippedClass?.BaseStrength ?? 0));

	public byte Magic => (byte)(Character.BaseMagic + this.Character.SpecialtyClass.BaseMagic + (this.EquippedClass?.BaseMagic ?? 0));

	public byte Defense =>
		(byte)(Character.BaseDefense + this.Character.SpecialtyClass.BaseDefense + (this.EquippedClass?.BaseDefense ?? 0));

	public byte Resistance =>
		(byte)(Character.BaseResistance + this.Character.SpecialtyClass.BaseResistance + (this.EquippedClass?.BaseResistance ?? 0));
}

File: client/scripts/session-character.js
class SessionCharacter {
    static baseHP = 10;
    static baseMP = 10;
    static baseStrength = 5;
    static baseMagic = 5;
    static baseDefense = 5;
    static baseResistance = 5;
    static maxHPPotions = 3;
    static maxMPPotions = 3;

    /**
     * 
     * @param {object} serverCharacter
     */
    constructor(serverCharacter) {
        this.full = serverCharacter.full;
        this.currentHP = serverCharacter.currentHP;
        this.hpPotions = serverCharacter.hpPotions;
        this.currentMP = serverCharacter.currentMP;
        this.mpPotions = serverCharacter.mpPotions;
        this.equippedClass = serverCharacter.equippedClass;
        this.buffs = serverCharacter.buffs;
        this.gems = serverCharacter.gems;
        this.tempStats = serverCharacter.tempStats;
    }

    /** @type {Character} */
    full;

    /** @type {number} */
    get maxHP() {
        return Math.max(5, SessionCharacter.baseHP +
            this.specialtyClass.hp +
            (this.equippedClass == null ? 0 : this.equippedClass.hp)) +
            (this.tempStats["maxHP"] ?? 0);
    }
    /** @type {number} */
    get maxMP() {
        return Math.max(5, SessionCharacter.baseMP +
            this.full.specialtyClass.mp +
            (this.equippedClass == null ? 0 : this.equippedClass.mp)) +
            (this.tempStats["maxMP"] ?? 0);
    }
    /** @type {number} */
    get strength() {
        return Math.max(0, SessionCharacter.baseStrength +
            this.full.specialtyClass.strength +
            (this.equippedClass == null ? 0 : this.equippedClass.strength)) +
            (this.tempStats["str"] ?? 0);
    }
    /** @type {number} */
    get magic() {
        return Math.max(0, SessionCharacter.baseMagic +
            this.full.specialtyClass.magic +
            (this.equippedClass == null ? 0 : this.equippedClass.magic)) +
            (this.tempStats["mag"] ?? 0);
    }
    /** @type {number} */
    get defense() {
        return Math.max(0, SessionCharacter.baseDefense +
            this.full.specialtyClass.defense +
            (this.equippedClass == null ? 0 : this.equippedClass.defense)) +
            (this.tempStats["def"] ?? 0);
    }
    /** @type {number} */
    get resistance() {
        return Math.max(0, SessionCharacter.baseResistance +
            this.full.specialtyClass.resistance +
            (this.equippedClass == null ? 0 : this.equippedClass.resistance)) +
            (this.tempStats["res"] ?? 0);
    }

    /** @type {string} */
    get name() { return this.full.name; }
    /** @type {PronounList} */
    get pronouns() { return this.full.pronouns; }
    /** @type {string} */
    get species() { return this.full.species; }
    /** @type {string} */
    get player() { return this.full.player; }
    /** @type {string} */
    get image() { return this.full.image; }
    /** @type {JobClass} */
    get specialtyClass() { return this.full.specialtyClass; }

    /** @type {number} */
    currentHP;
    /** @type {number} */
    hpPotions;
    /** @type {number} */
    currentMP;
    /** @type {number} */
    mpPotions;
    /** @type {JobClass} */
    equippedClass = null;
    /** @type {{ [key: string]: Buff }} */
    buffs = {};
    /** @type {Gem[]} */
    gems = [null, null, null, null, null, null, null];

    /** @type {{ [id: string]: number }} */
    tempStats = {};

    toJSON() {
        return { ...this.full, ...this };
    }
}

class Buff {
    name = "";
    description = "";
    icon = "";
    turnsRemaining = 0;
}

class Gem {
    name = "";
    description = "";
    color = "";
}
